st2common/st2common/models/api/base.py

```diff
@@ -71,7 +70,22 @@ def __json__(self):
         return vars(self)
 
     def validate(self):
-        VALIDATOR(getattr(self, 'schema', {})).validate(vars(self))
+        """
+        Perform validation and return cleaned object on success.
+        """
+        schema = getattr(self, 'schema', {})
+        attributes = vars(self)
+
+        cleaned = util_schema.validate(instance=attributes, schema=schema,
+                                       cls=util_schema.CustomValidator, use_default=True)
+
+        # Note: For backward compatibility reasons, we re-assign attributes one the current
```

Comment:
I would personally prefer a different interface where `validate()` doesn't mutate object in place but returns a new "cleaned" object?

That's similar to how most other existing validation frameworks (in Django, mongonegine, etc.) work.

Changing it would obviously require more work since all the existing code needs to be updating, but I think it's worth it. What do others think?